binding.cc

```diff
@@ -1467,163 +1456,141 @@ rocksdb_native__on_write(rocksdb_write_batch_t *handle, int status) {
-      assert(err == 0);
-    } else {
-      err = js_get_null(env, &error);
+      err = js_create_string(env, req->handle.error, error.emplace());
       assert(err == 0);
     }
 
-    js_value_t *ctx;
-    err = js_get_reference_value(env, req->ctx, &ctx);
-    assert(err == 0);
-
-    js_value_t *cb;
-    err = js_get_reference_value(env, req->on_status, &cb);
+    js_receiver_t ctx;
+    err = js_get_reference_value(env, req->ctx, ctx);
     assert(err == 0);
 
-    err = js_delete_reference(env, req->on_status);
+    cb_on_write_t cb;
+    err = js_get_reference_value(env, req->on_write, cb);
     assert(err == 0);
 
-    err = js_delete_reference(env, req->ctx);
-    assert(err == 0);
+    req->on_write.reset();
+    req->ctx.reset();
 
-    js_call_function_with_checkpoint(env, ctx, cb, 1, (js_value_t *[]) {error}, NULL);
+    js_call_function_with_checkpoint(env, cb, ctx, error);
 
     err = js_close_handle_scope(env, scope);
     assert(err == 0);
   }
 }
 
-static js_value_t *
-rocksdb_native_write(js_env_t *env, js_callback_info_t *info) {
+static void
+rocksdb_native_write(
+  js_env_t *env,
+  js_arraybuffer_span_of_t<rocksdb_native_t, 1> db,
+  js_arraybuffer_span_of_t<rocksdb_native_write_batch_t, 1> req,
+  js_array_t operations,
+  js_receiver_t ctx,
+  cb_on_write_t on_write
+) {
   int err;
 
-  size_t argc = 5;
-  js_value_t *argv[5];
-
-  err = js_get_callback_info(env, info, &argc, argv, NULL, NULL);
-  assert(err == 0);
-
-  assert(argc == 5);
-
-  rocksdb_native_t *db;
-  err = js_get_arraybuffer_info(env, argv[0], (void **) &db, NULL);
-  assert(err == 0);
-
-  rocksdb_native_write_batch_t *req;
-  err = js_get_arraybuffer_info(env, argv[1], (void **) &req, NULL);
+  err = js_create_reference(env, ctx, req->ctx);
   assert(err == 0);
 
-  uint32_t len;
-  err = js_get_array_length(env, argv[2], &len);
+  err = js_create_reference(env, on_write, req->on_write);
   assert(err == 0);
 
-  err = js_create_reference(env, argv[3], 1, &req->ctx);
-  assert(err == 0);
+  std::vector<js_object_t> elements;
 
-  err = js_create_reference(env, argv[4], 1, &req->on_status);
+  err = js_get_array_elements(env, operations, elements);
   assert(err == 0);
 
-  auto elements = reinterpret_cast<js_value_t **>(malloc(len * sizeof(js_value_t *)));
-
-  err = js_get_array_elements(env, argv[2], elements, len, 0, NULL);
-  assert(err == 0);
+  const auto len = elements.size();
 
   for (uint32_t i = 0; i < len; i++) {
-    js_value_t *write = elements[i];
-
-    js_value_t *property;
+    js_object_t write = elements[i];
 
-    err = js_get_named_property(env, write, "type", &property);
+    js_handle_t type_property;
+    err = js_get_property(env, write, "type", type_property);
     assert(err == 0);
 
     rocksdb_write_type_t type;
-    err = js_get_value_uint32(env, property, reinterpret_cast<uint32_t *>(&type));
+    err = js_get_value_uint32(env, static_cast<js_value_t *>(type_property), reinterpret_cast<uint32_t *>(&type));
```

Comment:
I'll look into a more ergonomic way to do this, we shouldn't have to escape to a raw ABI handle.

Reply:
```suggestion
    err = js_get_property(env, write, "type", reinterpret_cast<uint32_t &>(type));
```

That should actually be enough ☝️

Reply:
super